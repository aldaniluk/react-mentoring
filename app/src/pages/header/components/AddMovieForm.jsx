import styled from 'styled-components';
import genresFromJson from '@assets/data/genres'
import { FormComponent, FormOptionName, FormOptionInput, FormOptionDropdown, ColoredButton, TransparentButton } from '@globalComponents'
import { useState, useEffect } from 'react'

const Title = styled.div`
    color: white;
    font-size: 30px;
`

const ButtonContainer = styled.div`
    margin-top: 40px;
    text-align: right;
`

const ColoredButtonWrapper = styled(ColoredButton)`
    margin-left: 10px;
`

function AddMovieForm(props) {
    const [genres, setGenres] = useState([]);
    
    useEffect(() => {
        Promise
            .resolve(genresFromJson)
            .then(genres => setGenres(genres))
    }, [])

    return (
        <FormComponent>
            <Title>ADD MOVIE</Title>
            <FormOptionName>TITLE</FormOptionName>
            <FormOptionInput placeholder='Title here' />
            <FormOptionName>RELEASE DATE</FormOptionName>
            <FormOptionInput placeholder='Release date here' />
            <FormOptionName>MOVIE URL</FormOptionName>
            <FormOptionInput placeholder='Movie URL gere' />
            <FormOptionName>GENRE</FormOptionName>
            <FormOptionDropdown>
                {
                    genres.map(genre => (
                            <option key={genre.id}>
                                {genre.name}
                            </option>
                        )
                    )
                }
            </FormOptionDropdown>
            <FormOptionName>OVERVIEW</FormOptionName>
            <FormOptionInput placeholder='Overview here' />
            <FormOptionName>RUNTIME</FormOptionName>
            <FormOptionInput placeholder='Runtime here' />
            <ButtonContainer>
                <TransparentButton onClick={props.close}>RESET</TransparentButton>
                <ColoredButtonWrapper>SUBMIT</ColoredButtonWrapper>
            </ButtonContainer>
        </FormComponent>
    )
}

export { AddMovieForm }