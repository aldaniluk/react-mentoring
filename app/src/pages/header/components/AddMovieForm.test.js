import React from 'react';
import { AddMovieForm } from './AddMovieForm.jsx';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Movie } from '@models';

const checkInput = async (input, value, errorText) => {
    let onSubmitAddMovie = jest.fn();
    let onSubmitClose = jest.fn();

    render(<AddMovieForm addMovie={onSubmitAddMovie} close={onSubmitClose} />);

    if(input != 'title'){
        setInput('title', titleCorrectValue);
    } 
    else if(value) {
        setInput('title', value);
    }

    if(input != 'release_date'){
        setInput('release_date', releaseDateCorrectValue);
    } 
    else if(value) {
        setInput('release_date', value);
    }

    if(input != 'poster_path'){
        setInput('poster_path', posterPathCorrectValue);
    } 
    else if(value) {
        setInput('poster_path', value);
    }

    if(input != 'genres'){
        setInput('genres', genresCorrectValue);
    } 
    else if(value) {
        setInput('genres', value);
    }

    if(input != 'overview'){
        setInput('overview', overviewCorrectValue);
    } 
    else if(value) {
        setInput('overview', value);
    }

    if(input != 'runtime'){
        setInput('runtime', runtimeCorrectValue);
    } 
    else if(value) {
        setInput('runtime', value);
    }

    let submitButton = screen.queryByText('SUBMIT');

    fireEvent.click(submitButton);

    let error = await screen.findByText(errorText);
    expect(error).toBeInTheDocument();
    expect(onSubmitAddMovie).not.toBeCalled();
    expect(onSubmitClose).not.toBeCalled();
}

const titleCorrectValue = 'Some Movie';
const releaseDateCorrectValue = '2021-01-01';
const posterPathCorrectValue = 'http://some_url';
const genresCorrectValue = 'Crime,Horror';
const overviewCorrectValue = 'Some overview';
const runtimeCorrectValue = '123';

const setInput = (inputTitle, value) => {
    let input = screen.getByTitle(inputTitle);
    fireEvent.change(input, { target: { value: value } });
} 

describe('AddMovieForm', () => {
    it('AddMovieForm UI testing', () => {
        const { asFragment } = render(<AddMovieForm />);

        expect(asFragment()).toMatchSnapshot();
    });

    it('AddMovieForm happy path testing', async () => {
        let onSubmitAddMovie = jest.fn();
        let onSubmitClose = jest.fn();

        render(<AddMovieForm addMovie={onSubmitAddMovie} close={onSubmitClose} />);

        setInput('title', titleCorrectValue);
        setInput('release_date', releaseDateCorrectValue);
        setInput('poster_path', posterPathCorrectValue);
        setInput('genres', genresCorrectValue);
        setInput('overview', overviewCorrectValue);
        setInput('runtime', runtimeCorrectValue);

        let submitButton = screen.queryByText('SUBMIT');

        fireEvent.click(submitButton);

        let movie = new Movie(titleCorrectValue, releaseDateCorrectValue, posterPathCorrectValue, genresCorrectValue.split(','), overviewCorrectValue, Number.parseInt(runtimeCorrectValue));

        await waitFor(() => expect(onSubmitAddMovie).toHaveBeenCalledWith(movie));
        await waitFor(() => expect(onSubmitClose).toBeCalled());
    });

    it('AddMovieForm Title input testing: value is required', 
        () => checkInput('title', null, 'Title is required'));

    it('AddMovieForm Release date input testing: value is required', 
        () => checkInput('release_date', null, 'Release date is required'));

    it('AddMovieForm Release date input testing: value is incorrect', 
        () => checkInput('release_date', 'incorrect value', 'Release date must be in format yyyy-mm-dd'));

    it('AddMovieForm Url input testing: value is required', 
        () => checkInput('poster_path', null, 'Url is required'));

    it('AddMovieForm Url input testing: value is incorrect', 
        () => checkInput('poster_path', 'incorrect value', 'Url is incorrect'));

    it('AddMovieForm Genres input testing: value is required', 
        () => checkInput('genres', null, 'Genres are required'));

    it('AddMovieForm Genres input testing: value is incorrect', 
        () => checkInput('genres', 'Crime,Horro1', 'Some of genres are incorrect'));

    it('AddMovieForm Overview input testing: value is required', 
        () => checkInput('overview', null, 'Overview is required'));

    it('AddMovieForm Runtime input testing: value is required', 
        () => checkInput('runtime', null, 'Runtime is required'));

    it('AddMovieForm Runtime input testing: value is incorrect (!=integer)', 
        () => checkInput('runtime', 'text', 'Runtime must be an integer'));

    it('AddMovieForm Runtime input testing: value is incorrect (==0)', 
        () => checkInput('runtime', '0', 'Runtime must be greater that 0'));
})