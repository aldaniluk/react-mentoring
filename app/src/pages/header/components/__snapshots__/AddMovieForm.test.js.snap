// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMovieForm AddMovieForm UI testing 1`] = `
<DocumentFragment>
  <div
    class="sc-eCssSg cqcqLB"
  >
    <form
      class="sc-jSgupP QSFhL"
    >
      <div
        class="sc-khAkjo iZFSjE"
      >
        ADD MOVIE
      </div>
      <div
        class="sc-gKsewC eXvkrr"
      >
        TITLE
      </div>
      <input
        class="sc-iBPRYJ ddlvwJ"
        id="title"
        name="title"
        placeholder="Title here"
        title="title"
        value=""
      />
      <div
        class="sc-gKsewC eXvkrr"
      >
        RELEASE DATE
      </div>
      <input
        class="sc-iBPRYJ ddlvwJ"
        id="release_date"
        name="release_date"
        placeholder="Release date here"
        title="release_date"
        value=""
      />
      <div
        class="sc-gKsewC eXvkrr"
      >
        MOVIE URL
      </div>
      <input
        class="sc-iBPRYJ ddlvwJ"
        id="poster_path"
        name="poster_path"
        placeholder="Url here"
        title="poster_path"
        value=""
      />
      <div
        class="sc-gKsewC eXvkrr"
      >
        GENRES
      </div>
      <input
        class="sc-iBPRYJ ddlvwJ"
        id="genres"
        name="genres"
        placeholder="Genres here"
        title="genres"
        value=""
      />
      <div
        class="sc-gKsewC eXvkrr"
      >
        OVERVIEW
      </div>
      <input
        class="sc-iBPRYJ ddlvwJ"
        id="overview"
        name="overview"
        placeholder="Overview here"
        title="overview"
        value=""
      />
      <div
        class="sc-gKsewC eXvkrr"
      >
        RUNTIME
      </div>
      <input
        class="sc-iBPRYJ ddlvwJ"
        id="runtime"
        name="runtime"
        placeholder="Runtime here"
        title="runtime"
        value=""
      />
      <div
        class="sc-hTZhsR Kjldg"
      >
        <button
          class="sc-pFZIQ hJcOwJ"
        >
          RESET
        </button>
        <button
          class="sc-jrAGrp sc-jgHCyG jccncV idCPzI"
          type="submit"
        >
          SUBMIT
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
